### PR TITLE
Bluetooth: Host: Fix SMP issues

### LINKED ISSUE #59788: `bt_unpair()` call in security changed callback causes assertion in `keys.c`
**Describe the bug**
Assert is triggered in the `smp_pairing_complete()` context after application removes the bonding information with the `bt_unpair()` API in the context of security changed callback.
The stack tries updating the keys usage database with the internal `bt_keys_store()` to finalize the Bluetooth pairing flow (`smp_pairing_complete()`). 
As the pairing keys are removed during the `bt_unpair()` API call, this operation results in the assert as the keys structure pointer is NULL.

**To Reproduce**
1. Modify `samples/bluetooth/peripheral_hids` by commenting out `bt_conn_set_security()` and adding `bt_unpair()` in `security_changed()` callback:
```
diff --git a/samples/bluetooth/peripheral_hids/src/main.c b/samples/bluetooth/peripheral_hids/src/main.c
index 13109d40d9..2ecbd65a37 100644
--- a/samples/bluetooth/peripheral_hids/src/main.c
+++ b/samples/bluetooth/peripheral_hids/src/main.c
@@ -44,9 +44,9 @@ static void connected(struct bt_conn *conn, uint8_t err)
 
 	printk("Connected %s\n", addr);
 
-	if (bt_conn_set_security(conn, BT_SECURITY_L2)) {
-		printk("Failed to set security\n");
-	}
+	// if (bt_conn_set_security(conn, BT_SECURITY_L2)) {
+	// 	printk("Failed to set security\n");
+	// }
 }
 
 static void disconnected(struct bt_conn *conn, uint8_t reason)
@@ -71,6 +71,11 @@ static void security_changed(struct bt_conn *conn, bt_security_t level,
 		printk("Security failed: %s level %u err %d\n", addr, level,
 		       err);
 	}
+
+	printk("bt_unpair start\n");
+	int err2 = bt_unpair(BT_ID_DEFAULT, bt_conn_get_dst(conn));
+	printk("bt_unpair end, return: %d\n", err2);
+
 }
 
 BT_CONN_CB_DEFINE(conn_callbacks) = {

```

2. Build:
```
west build --pristine -d build -b nrf52840dk_nrf52840 -- -DCONFIG_BT_SMP_LOG_LEVEL_DBG=y -DCONFIG_BT_KEYS_LOG_LEVEL_DBG=y -DCONFIG_ASSERT=y -DCONFIG_SAMPLE_BT_USE_AUTHENTICATION=n -DCONFIG_LOG_PROCESS_THREAD_STACK_SIZE=8192
```

3. Flash:
```
west flash --skip-rebuild --erase -d build
```

4. Bond with the device (`Test HoG mouse`) using `nRF Connect for Mobile` or system level pairing.

**Expected behavior**
Host desists from keys usage database update using `bt_keys_store()` when there is nothing to be stored (`keys == NULL`).

**Logs:**
```
*** Booting Zephyr OS build zephyr-v3.4.0-433-g3be68bd4fe48 ***
[00:00:00.267,974] <inf> fs_nvs: 8 Sectors of 4096 bytes
[00:00:00.268,005] <inf> fs_nvs: alloc wra: 0, fe8
[00:00:00.268,005] <inf> fs_nvs: data wra: 0, 0
[00:00:00.270,568] <inf> bt_hci_core: HW Platform: Nordic Semiconductor (0x0002)
[00:00:00.270,599] <inf> bt_hci_core: HW Variant: nRF52x (0x0002)
[00:00:00.270,629] <inf> bt_hci_core: Firmware: Standard Bluetooth controller (0x00) Version 3.4 Build 99
[00:00:00.271,057] <dbg> bt_smp: bt_smp_init: LE SC enabled
[00:00:00.271,240] <inf> bt_hci_core: No ID address. App must call settings_load()
Bluetooth initialized
[00:00:00.271,759] <inf> bt_hci_core: Identity: D1:C2:0E:ED:76:AB (random)
[00:00:00.271,789] <inf> bt_hci_core: HCI: version 5.4 (0x0d) revision 0x0000, manufacturer 0x05f1
[00:00:00.271,820] <inf> bt_hci_core: LMP: version 5.4 (0x0d) subver 0xffff
Advertising successfully started
[00:00:00.542,724] <dbg> bt_smp: bt_smp_pkey_ready: 
[00:00:14.026,031] <dbg> bt_keys: bt_keys_find_irk: 40:FC:FE:F3:A0:B8 (random)
[00:00:14.026,214] <dbg> bt_keys: bt_keys_find_irk: No IRK for 40:FC:FE:F3:A0:B8 (random)
[00:00:14.026,275] <dbg> bt_smp: bt_smp_accept: conn 0x20001238 handle 0
[00:00:14.026,275] <dbg> bt_smp: bt_smp_connected: chan 0x20001504 cid 0x0006
[00:00:14.026,519] <dbg> bt_keys: bt_keys_find_addr: 40:FC:FE:F3:A0:B8 (random)
Connected 40:FC:FE:F3:A0:B8 (random)
[00:00:14.233,337] <dbg> bt_smp: bt_smp_recv: Received SMP code 0x01 len 6
[00:00:14.233,367] <dbg> bt_smp: smp_pairing_req: req: io_capability 0x04, oob_flag 0x00, auth_req 0x2D, max_key_size 0x10, init_key_dist 0x0F, resp_key_dist 0x0F
[00:00:14.233,551] <dbg> bt_keys: bt_keys_get_addr: 40:FC:FE:F3:A0:B8 (random)
[00:00:14.233,764] <dbg> bt_keys: bt_keys_get_addr: created 0x20001c4c for 40:FC:FE:F3:A0:B8 (random)
[00:00:14.234,985] <dbg> bt_smp: smp_init: prnd ea45b8b30d1bdfcd7e769d65e70ad9d2
[00:00:14.235,015] <dbg> bt_smp: smp_pairing_req: rsp: io_capability 0x03, oob_flag 0x00, auth_req 0x09, max_key_size 0x10, init_key_dist 0x02, resp_key_dist 0x00
[00:00:14.332,214] <dbg> bt_smp: bt_smp_recv: Received SMP code 0x0c len 64
[00:00:14.332,244] <dbg> bt_smp: smp_public_key: 
[00:00:14.428,466] <dbg> bt_smp: bt_smp_recv: Received SMP code 0x04 len 16
[00:00:14.428,497] <dbg> bt_smp: smp_pairing_random: 
[00:00:14.592,376] <dbg> bt_smp: bt_smp_dhkey_ready: 0x20008414
[00:00:16.963,226] <dbg> bt_smp: bt_smp_recv: Received SMP code 0x0d len 16
[00:00:16.963,256] <dbg> bt_smp: smp_dhkey_check: 
[00:00:16.968,658] <dbg> bt_smp: sc_smp_send_dhkey_check: 
[00:00:17.694,580] <dbg> bt_keys: bt_keys_clear: 40:FC:FE:F3:A0:B8 (random) (keys 0x0000)
[00:00:17.695,037] <dbg> bt_keys: bt_keys_get_addr: 40:FC:FE:F3:A0:B8 (random)
[00:00:17.695,220] <dbg> bt_keys: bt_keys_get_addr: created 0x20001c4c for 40:FC:FE:F3:A0:B8 (random)
[00:00:17.695,251] <dbg> bt_smp: bt_smp_encrypt_change: chan 0x20001504 conn 0x20001238 handle 0 encrypt 0x01 hci status 0x00
Security changed: 40:FC:FE:F3:A0:B8 (random) level 2
bt_unpair start
[00:00:17.695,861] <dbg> bt_keys: bt_keys_clear: 40:FC:FE:F3:A0:B8 (random) (keys 0x0020)
bt_unpair returned: 0
[00:00:17.792,022] <dbg> bt_smp: smp_ident_info: 
--- 1 messages dropped ---
[00:00:17.792,205] <dbg> bt_keys: bt_keys_get_type: type 2 40:FC:FE:F3:A0:B8 (random)
[00:00:17.792,449] <dbg> bt_keys: bt_keys_find: type 2 40:FC:FE:F3:A0:B8 (random)
[00:00:17.792,633] <dbg> bt_keys: bt_keys_get_addr: 40:FC:FE:F3:A0:B8 (random)
[00:00:17.792,877] <dbg> bt_keys: bt_keys_get_addr: created 0x20001c4c for 40:FC:FE:F3:A0:B8 (random)
[00:00:17.792,999] <dbg> bt_smp: bt_smp_recv: Received SMP code 0x09 len 7
[00:00:17.793,212] <dbg> bt_smp: smp_ident_addr_info: identity 6C:F7:84:9A:F9:CB (public)
[00:00:17.793,487] <dbg> bt_keys: bt_keys_find_addr: 6C:F7:84:9A:F9:CB (public)
[00:00:17.793,670] <dbg> bt_keys: bt_keys_get_type: type 2 40:FC:FE:F3:A0:B8 (random)
[00:00:17.793,853] <dbg> bt_keys: bt_keys_find: type 2 40:FC:FE:F3:A0:B8 (random)
[00:00:17.794,036] <dbg> bt_keys: bt_keys_find_addr: 6C:F7:84:9A:F9:CB (public)
[00:00:17.794,860] <dbg> bt_smp: smp_pairing_complete: status 0x0
ASSERTION FAIL [keys != ((void *)0)] @ WEST_TOPDIR/zephyr/subsys/bluetooth/host/keys.c:326
[00:00:17.794,921] <err> os: r0/a1:  0x00000004  r1/a2:  0x00000146  r2/a3:  0x0000000e
[00:00:17.794,921] <err> os: r3/a4:  0x20000c58 r12/ip:  0x200069e0 r14/lr:  0x0001116d
[00:00:17.794,952] <err> os:  xpsr:  0x61000000
[00:00:17.794,952] <err> os: Faulting instruction address (r15/pc): 0x00027f7a
[00:00:17.795,013] <err> os: >>> ZEPHYR FATAL ERROR 4: Kernel panic on CPU 0
[00:00:17.795,043] <err> os: Current thread: 0x20001570 (unknown)
[00:00:17.969,390] <err> os: Halting system
```

**Environment:**
 - Ubuntu 22.04 LTS
 - Zephyr v3.4.0, commit: 3be68bd4fe48d4a3723ada18af9466bb646a5ad6
 - Tested on `nrf52840dk_nrf52840`
 - Mobile: iOS (iPhone XR, iOS 16.5); Android (Redmi Note 10S, Android 11)
 
**Additional context**
It seems that BR counterpart (`smp_pairing_br_complete()`) does perform check before calling `bt_keys_store()`:
```
if (bond_flag && keys) {
	bt_keys_store(keys);
}
```
However, in `smp_pairing_complete()` there is no such check:
```
if (bond_flag) {
	bt_keys_store(conn->le.keys);
}
```

### LINKED ISSUE #61465: A disconnect shortly after a pairing completes results in pairing failed
**Describe the bug**

A device that disconnects the BLE connection shortly after having completed a pairing will for some controllers make the Zephyr host report the pairing as failed.

This happens because the smp_pairing_complete handler in smp.c detects that the connection is in the BT_CONN_DISCONNECT_COMPLETE state and then aborts the pairing.

This is because the disconnect event is processed before the smp_pairing_complete handler is called even though the pairing completes before the disconnect in the air.

**To Reproduce**

#### Devices

1. Device 1 - A BLE device that wants to pair with Device 2.
1. Device 2 - A BLE device running Zephyr version 3.4.0.

#### Steps to Reproduce

1. Device 2 is in pairing mode and advertises.
1. Device 1 connects to Device 2.
1. Device 1 writes a value to a GATT characteristic. (Not sure this step is necessary to reproduce the bug)
1. Device 1 pairs with Device 2.
1. Pairing completes successfully in the air.
1. Device 1 disconnects from Device 2 very shortly after pairing completed in the air.
1. Pairing is reported failed to the application in Device 2.

#### Analysis

The order of events received by the application in device 2 depends on the choice of bluetooth controller.

The following order of events is seen in the debug output when using PacketCraft controller version 3349 for the nRF5340 Audio board as Device 2.

| Prefix      | Description               |
|-------------|---------------------------|
| APP         | The application           |
| ZEPHYR CONN | The conn module in Zephyr |
| ZEPHYR SMP  | The smp module in Zephyr  |

1. APP: connected handler
1. ZEPHYR SMP: smp_pairing_req
1. APP: pairing accept handler
1. ZEPHYR SMP: legacy_pairing_req and more SMP stuff from the pairing process.
1. APP: settings save handler (remove key)
1. ZEPHYR SMP: bt_smp_encrypt_change and more SMP stuff from the pairing process.
1. APP: security changed handler
1. APP: identity resolved
1. APP: settings save handler (remove ccc)
1. ZEPHYR CONN: disconnect complete (state set)
1. ZEPHYR SMP: smp_signing_info
1. ZEPHYR SMP: smp_pairing_complete - Not connected!
1. APP: settings save handler (remove key)
1. APP: pairing failed
1. ZEPHYR SMP: bt_smp_disconnected
1. APP: disconnected handler

A different order of events can occur with other controllers. The Zephyr controller gives a different result on the nRF5340 Audio board. But the same pairing failed event also happens with our own Zephyr based controller for our own platform.

The above order of events observed shows that the disconnect state is set before the smp_pairing_complete handler is called which makes it abort pairing even though the pairing DID complete successfully.

**Expected behavior**

The diconnect event should be handled after the pairing complete handler has been called since it comes after the pairing events (SMP) in the air time-wise.

**Impact**

This bug makes it impossible for a device to only maintain a connection while it is using it. For instance as described in the steps to reproduce.

**Logs and console output**

None.

**Environment (please complete the following information):**

- Zephyr 3.4.0 (Based on 13a6c454529e64cdb9f69cab0f0a9fb8c539b192)

**Additional context**

None.

### PR DESCRIPTION
**Pull request by theob-pro**
Add a test to reproduce #59788 and #61465.

Fixes those two issues, if the bonding information has been cleared before pairing had a chance to complete (probably by the application), indicate this by setting an appropriate log message.

Also check that keys exist before calling `bt_keys_store`.

Fixes #59788 #61465

### CODE REVIEW

**Review by jori-nordic — DISMISSED** [review]

**Review by jhedberg — DISMISSED** [review]

File: tests/bsim/bluetooth/host/security/unpairing_security_changed/src/bs_bt_utils.c [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+
+void scan_connect_to_first_result(void)
+{
+	int err;
+
+	err = bt_le_scan_start(BT_LE_SCAN_PASSIVE, scan_connect_to_first_result__device_found);
+	ASSERT(!err, "Err bt_le_scan_start %d", err);
+}
+
+void disconnect(void)
```
Comment (jilu-ot) [inline tests/bsim/bluetooth/host/security/unpairing_security_changed/src/bs_bt_utils.c:171]: Looks like disconnect() is not called.

File: tests/bsim/bluetooth/host/security/unpairing_security_changed/src/bs_bt_utils.c [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+	err = bt_conn_auth_info_cb_register(&bt_conn_auth_info_cb);
+	ASSERT(!err, "bt_conn_auth_info_cb_register failed.\n");
+
+	err = settings_load();
+	if (err) {
+		FAIL("Settings load failed (err %d)\n", err);
+	}
+}
+
+static void scan_connect_to_first_result__device_found(const bt_addr_le_t *addr, int8_t rssi,
```
Comment (jilu-ot) [inline tests/bsim/bluetooth/host/security/unpairing_security_changed/src/bs_bt_utils.c:138]: Scan is not started here. Name is misleading.

File: tests/bsim/bluetooth/host/security/unpairing_security_changed/src/central.c [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+#include <zephyr/bluetooth/conn.h>
+#include <zephyr/bluetooth/bluetooth.h>
+
+#include "bs_bt_utils.h"
+
+LOG_MODULE_REGISTER(test_central, LOG_LEVEL_DBG);
+
+BUILD_ASSERT(CONFIG_BT_BONDABLE, "CONFIG_BT_BONDABLE must be enabled by default.");
+
+void central(void)
```
Comment (jilu-ot) [inline tests/bsim/bluetooth/host/security/unpairing_security_changed/src/central.c:20]: Where does the central disconnect after pairing has completed?

Comment (jilu-ot) [reply tests/bsim/bluetooth/host/security/unpairing_security_changed/src/central.c:20]: Remember it is the central that disconnects right after pairing completes in the description of issue #61465.

Comment (theob-pro) [reply tests/bsim/bluetooth/host/security/unpairing_security_changed/src/central.c:20]: The behavior of the Peripheral is the same as the Central, as far as the `pairing_complete` callback is concerned. Have you tested this patch on your platform? We think that it does fix the issue.

We can't test the Central in the same way, as the Central will call the security changed callback _before_ distributing the keys: the Central will not send its keys until it has received the Peripheral's.

File: tests/bsim/bluetooth/host/security/unpairing_security_changed/src/bs_bt_utils.c (hunk 2 above)
Comment (theob-pro) [reply tests/bsim/bluetooth/host/security/unpairing_security_changed/src/bs_bt_utils.c:138]: This is the callback of `bt_le_scan_start`.

**Comment by theob-pro** [comment]
Rebased to update test and split it into two different runs.

File: tests/bsim/bluetooth/host/security/security_changed_callback/src/bs_bt_utils.c [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+DEFINE_FLAG(flag_pairing_complete);
+DEFINE_FLAG(flag_bonded);
+DEFINE_FLAG(flag_not_bonded);
+
+static void pairing_complete(struct bt_conn *conn, bool bonded)
+{
+	LOG_ERR("pairing complete");
+	SET_FLAG(flag_pairing_complete);
+
+	if (bonded) {
```
Comment (jori-nordic) [inline tests/bsim/bluetooth/host/security/security_changed_callback/src/bs_bt_utils.c:98]: Can merge the two ifs

File: tests/bsim/bluetooth/host/security/security_changed_callback/src/bs_bt_utils.c [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+	}
+
+	if (bonded) {
+		SET_FLAG(flag_bonded);
+	} else {
+		SET_FLAG(flag_not_bonded);
+	}
+}
+
+DEFINE_FLAG(flag_pairing_failed);
```
Comment (jori-nordic) [inline tests/bsim/bluetooth/host/security/security_changed_callback/src/bs_bt_utils.c:111]: we don't seem to use that one

File: tests/bsim/bluetooth/host/security/security_changed_callback/src/bs_bt_utils.c [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+{
+	char addr_str[BT_ADDR_LE_STR_LEN];
+	int err;
+
+	if (g_conn != NULL) {
+		return;
+	}
+
+	/* We're only interested in connectable events */
+	if (type != BT_HCI_ADV_IND && type != BT_HCI_ADV_DIRECT_IND) {
```
Comment (jori-nordic) [inline tests/bsim/bluetooth/host/security/security_changed_callback/src/bs_bt_utils.c:143]: this doesn't really check that the evt is connectable. 
IMO we know what the peer is going to send, no need to validate, you can just remove this whole `if-else`.

File: tests/bsim/bluetooth/host/security/security_changed_callback/src/bs_bt_utils.c [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+	err = bt_conn_auth_info_cb_register(&bt_conn_auth_info_cb);
+	ASSERT(!err, "bt_conn_auth_info_cb_register failed.\n");
+
+	err = settings_load();
+	if (err) {
+		FAIL("Settings load failed (err %d)\n", err);
+	}
+}
+
+static void scan_cb_connect_to_first_result__device_found(const bt_addr_le_t *addr, int8_t rssi,
```
Comment (jori-nordic) [inline tests/bsim/bluetooth/host/security/security_changed_callback/src/bs_bt_utils.c:132]: can probably rename to `stop_scan_and_connect()`

File: tests/bsim/bluetooth/host/security/security_changed_callback/src/bs_bt_utils.h [hunk 8]
```diff
@@ -0,0 +1,79 @@
+/**
+ * Common functions and helpers for BSIM GATT tests
```
Comment (jori-nordic) [inline tests/bsim/bluetooth/host/security/security_changed_callback/src/bs_bt_utils.h:2]: update

File: tests/bsim/bluetooth/host/security/security_changed_callback/src/peripheral.c [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+	ASSERT(!err, "bt_conn_auth_info_cb_register failed.\n");
+
+	/* Disconnect in security changed callback */
+
+	advertise_connectable(BT_ID_DEFAULT, NULL);
+	wait_connected();
+
+	wait_disconnected();
+
+	clear_g_conn();
```
Comment (jori-nordic) [inline tests/bsim/bluetooth/host/security/security_changed_callback/src/peripheral.c:105]: are you using `g_conn` outside of `bt_bt_utils.c`? 
If you're not, you can remove it from the `.h` and also remove the `clear_g_conn()`.

If you use it, I would rather have `struct bt_conn *conn = wait_connected()` as the interface.

File: tests/bsim/bluetooth/host/security/security_changed_callback/src/peripheral.c [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+	FAIL("Pairing failed\n");
+}
+
+static void peripheral_security_changed_disconnect(struct bt_conn *conn,
+						   bt_security_t level,
+						   enum bt_security_err err)
+{
+	/* Try to trigger fault here */
+	k_msleep(2000);
+	LOG_ERR("do bad");
```
Comment (jori-nordic) [inline tests/bsim/bluetooth/host/security/security_changed_callback/src/peripheral.c:77]: LOG_INF, same for the other one

File: tests/bsim/bluetooth/host/security/security_changed_callback/src/peripheral.c [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+	peripheral_cb.security_changed = peripheral_security_changed_disconnect;
+	peripheral_auth_info_cb.pairing_failed = pairing_failed_disconnect;
+
+	bs_bt_utils_setup();
+
+	bt_conn_cb_register(&peripheral_cb);
+	err = bt_conn_auth_info_cb_register(&peripheral_auth_info_cb);
+	ASSERT(!err, "bt_conn_auth_info_cb_register failed.\n");
+
+	/* Disconnect in security changed callback */
```
Comment (jori-nordic) [inline tests/bsim/bluetooth/host/security/security_changed_callback/src/peripheral.c:98]: This comment should be above L89. same comment for the other testcase

File: tests/bsim/bluetooth/host/security/security_changed_callback/test_scripts/_env.sh [hunk 12]
```diff
@@ -0,0 +1,15 @@
+#!/bin/env bash
+# Copyright 2023 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+set -eu
+bash_source_dir="$(realpath "$(dirname "${BASH_SOURCE[0]}")")"
+
+: "${BSIM_OUT_PATH:?BSIM_OUT_PATH must be defined}"
+
+test_name="$(basename "$(realpath "$bash_source_dir/..")")"
```
Comment (jori-nordic) [inline tests/bsim/bluetooth/host/security/security_changed_callback/test_scripts/_env.sh:10]: doesn't seem to be used a lot. I think you can just remove this file

File: tests/bsim/bluetooth/host/security/security_changed_callback/test_scripts/security_changed_callback.sh [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+wait_for_background_jobs
+
+Execute "./${test_exe}" \
+  -v=${verbosity_level} -s=${simulation_id} -d=0 -testid=central
+
+Execute "./${test_exe}" \
+  -v=${verbosity_level} -s=${simulation_id} -d=1 -testid=peripheral_unpair_in_sec_cb
+
+Execute ./bs_2G4_phy_v1 -v=${verbosity_level} -s=${simulation_id} \
+  -D=2 -sim_length=60e6
```
Comment (jori-nordic) [inline tests/bsim/bluetooth/host/security/security_changed_callback/test_scripts/security_changed_callback.sh:33]: you can shorten that to save on precious CI credits :)

**Review by jori-nordic — CHANGES_REQUESTED** [review]

File: tests/bsim/bluetooth/host/security/unpairing_security_changed/src/central.c (hunk 3 above)
Comment (jilu-ot) [reply tests/bsim/bluetooth/host/security/unpairing_security_changed/src/central.c]: Ok. It just makes it a more unclear that the test is testing the steps to reproduce in the issue description. But ok, if it creates the same situation.

I tested the smp.c patch on our platform, and it works :-)

**Review by jori-nordic — DISMISSED** [review]

**Review by alwa-nordic — DISMISSED** [review]

File: tests/bsim/bluetooth/host/security/security_changed_callback/src/peripheral.c [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+	FAIL("Pairing succeed\n");
+}
+
+static void peripheral_security_changed_unpair(struct bt_conn *conn,
+					bt_security_t level,
+					enum bt_security_err err)
+{
+	/* Try to trigger fault here */
+	k_msleep(2000);
+	LOG_INF("do bad");
```
Comment (jilu-ot) [inline tests/bsim/bluetooth/host/security/security_changed_callback/src/peripheral.c:32]: Maybe change do bad to: "Remove pairing". For clarity.

File: tests/bsim/bluetooth/host/security/security_changed_callback/src/peripheral.c [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+	FAIL("Pairing failed\n");
+}
+
+static void peripheral_security_changed_disconnect(struct bt_conn *conn,
+						   bt_security_t level,
+						   enum bt_security_err err)
+{
+	/* Try to trigger fault here */
+	k_msleep(2000);
+	LOG_INF("do bad");
```
Comment (jilu-ot) [inline tests/bsim/bluetooth/host/security/security_changed_callback/src/peripheral.c:77]: Maybe change the log output to: "Disconnecting central", for clarity.

**Review by jilu-ot — APPROVED** [review]

**Review by jori-nordic — DISMISSED** [review]

**Comment by theob-pro** [comment]
Rebased on main to fix merge conflict.

**Review by jhedberg — APPROVED** [review]

**Review by alwa-nordic — APPROVED** [review]

**Pull request merged**